lib/src/code_generator/type.dart

```diff
@@ -47,6 +47,12 @@ abstract class Type {
   /// only. This string should not be printed as generated code.
   @override
   String toString();
+
+  /// Cache key used in various places to dedupe Types. By default this is just
+  /// the hash of the Type, but in many cases this does not dedupe sufficiently.
+  /// So Types that may be duplicated should override this to return a more
+  /// specific key. Types that are already deduped don't need to override this.
+  String cacheKey() => hashCode.toRadixString(36);
```

Comment:
Maybe document why `toString` as cacheKey doesn't work.

(I presume we could have name collisions?)

Reply:
Yeah. toString was just designed for debugging, so I haven't thought carefully about making it unambiguous.